### DISCUSSION TITLE
Transparent request retry - keep showing loading state?

### DISCUSSION THREAD

**Discussion by fabb** [bigskysoftware/htmx#2259] [Q&A]
I have this request flow in a non-htmx application, and wonder if it is possible to implement with htmx too.

Requests with an expired access token return a 401. I'd like to handle this 401 transparently for the user, refresh the user's access token and retry the original request - without showing an error for the initial request.

I've looked at the htmx events like `htmx:responseError` that would allow to retry requests as described here: https://github.com/bigskysoftware/htmx/discussions/1746#discussioncomment-6859968

But I want htmx to skip error handling for the first 401 response and keep showing the loading state. Is that somehow possible?

```mermaid
---
title: 401 Handling on the Client Side
---
flowchart TD
    A[Start] --> |send api request\ne.g. submit form| B{Response}
    B -->|200| Z1[Apply response]
    B -->|other| Z2[Show error]
    B --->|401| C{Success?}
    subgraph Background Auth Flow
        C[Try background login]-->E{Success?}
        E -->|no| G[Redirect to Login page]
        E -->|yes| F[Set-Cookie of new auth cookies]
    end
    G --> Z3[Original api request is NOT retried]
    F --> I[Retry original API request]
    I --> J{Response}
    J -->|200| Z5[Apply response]
    J -->|other| Z6[Show error]
    J -->|401| Z7[Show error]
```

**Comment by Telroshan** [comment]
Hey, I'm afraid it's not natively supported at the moment.
It seems what you want to do here is simply preventing htmx from removing the loading state when stumbling upon this specific error.
Looking at the lib's code, I can think of a pretty dirty workaround, but you'd have to try if it really works: 
- Bind a listener to [`htmx:beforeOnLoad`](https://htmx.org/events/#htmx:beforeOnLoad)
- Throw an exception lmao so that the [removeRequestIndicators call](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L3313) that is normally called right after the handler doesn't happen as it'll raise an error that will be caught by the surrounding `try` block
- You'll get another `htmx:onLoadError` event fired at this moment, but you just have do not do anything with it and your user wouldn't notice anyway
Pretty dirty I know but hey, it may work 🤷 (again, I haven't tested it so you'd have to try it for yourself to see if it really works as expected)

**Reply by fabb** [reply]
Interesting, thanks!
